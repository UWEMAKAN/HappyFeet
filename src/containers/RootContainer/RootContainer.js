/* eslint-disable import/no-named-as-default */
import React, { useState, useEffect } from 'react';
import { Route, Switch, Redirect } from 'react-router-dom';
import Spinner from '../../components/Common/Spinner/Spinner';
import HomePage from '../home/HomePage';
import MyShoesPage from '../myshoes/MyShoesPage';
import PageNotFound from '../../components/PageNotFound/PageNotFound';
import PaymentsPage from '../payment/PaymentsPage';
import { getShoes } from '../../api/shoesApi';

export const buyShoeHandler = (shoes, setShoe) => (id) => {
  const shoetoBuy = shoes.filter((s) => s.id === id)[0];
  setShoe(shoetoBuy);
};

export const myShoesHandler = (myShoes, setMyShoes, setShoe) => (myShoe) => {
  setMyShoes([...myShoes, myShoe]);
  setShoe({});
};

const RootContainer = () => {
  const [shoes, setShoes] = useState([]);
  const [shoe, setShoe] = useState({});
  const [myShoes, setMyShoes] = useState([]);

  useEffect(() => {
    if (!shoes.length) {
      getShoes()
        .then((fetchedShoes) => {
          setImmediate(() => setShoes(fetchedShoes));
          // setShoes(fetchedShoes);
        })
        .catch((err) => {
          // eslint-disable-next-line no-alert
          alert(`Loading shoes failed ${err}`);
        });
    }
  }, []);

  const buyShoe = buyShoeHandler(shoes, setShoe);
  const addToMyShoes = myShoesHandler(myShoes, setMyShoes, setShoe);

  return !shoes.length ? (
    <Spinner />
  ) : (
    <Switch>
      <Route exact path="/" component={() => (<HomePage shoes={shoes} buyShoe={buyShoe} />)} />
      <Route exact path="/shoes">
        <Redirect to="/" />
      </Route>
      <Route path="/myshoes" component={() => (<MyShoesPage myShoes={myShoes} />)} />
      <Route path="/payments" component={() => (<PaymentsPage shoe={shoe} addToMyShoes={addToMyShoes} />)} />
      <Route component={PageNotFound} />
    </Switch>
  );
};

export default RootContainer;
