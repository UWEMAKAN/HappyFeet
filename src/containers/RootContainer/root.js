/* eslint-disable import/no-named-as-default */
import React, { Component } from 'react';
import { Route, Switch, Redirect } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import propTypes from 'prop-types';
import * as shoesActions from '../../redux/actions/shoesActions/shoesActions';
import Spinner from '../../components/Common/Spinner/Spinner';
import HomePage from '../home/HomePage';
import MyShoesPage from '../myshoes/MyShoesPage';
import PageNotFound from '../../components/PageNotFound/PageNotFound';
import PaymentsPage from '../payment/PaymentsPage';

export class RootContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      shoes: [],
      shoe: {},
      myShoes: []
    };
  }

  componentDidMount() {
    const { shoes, actions } = this.props;
    if (!shoes.length) {
      actions.loadShoes()
        .then(() => {
          this.setState((prevState) => ({
            // eslint-disable-next-line react/destructuring-assignment
            ...prevState, shoes: this.props.shoes
          }));
        })
        .catch((err) => {
          // eslint-disable-next-line no-alert
          alert(`Loading shoes failed ${err}`);
        });
    }
  }

  buyShoe = (id) => {
    const { shoes } = this.state;
    const shoe = shoes.filter((s) => s.id === id)[0];
    this.setState((prevState) => ({
      ...prevState, shoe
    }));
  }

  addToMyShoes = (shoe) => {
    this.setState((prevState) => ({
      ...prevState,
      myShoes: [...prevState.myShoes, shoe],
      shoe: {}
    }));
  }

  render() {
    const { shoes, shoe, myShoes } = this.state;
    const { loading } = this.props;
    return !shoes.length || loading ? (
      <Spinner />
    ) : (
      <Switch>
        <Route exact path="/" component={() => (<HomePage shoes={shoes} buyShoe={this.buyShoe} />)} />
        <Route exact path="/shoes">
          <Redirect to="/" />
        </Route>
        <Route path="/myshoes" component={() => (<MyShoesPage myShoes={myShoes} />)} />
        <Route path="/payments" component={() => (<PaymentsPage shoe={shoe} addToMyShoes={this.addToMyShoes} />)} />
        <Route component={PageNotFound} />
      </Switch>
    );
  }
}

RootContainer.propTypes = {
  shoes: propTypes.instanceOf(Array).isRequired,
  actions: propTypes.instanceOf(Object).isRequired,
  loading: propTypes.bool.isRequired
};

const mapStateToProps = (state) => ({
  shoes: state.shoes,
  loading: state.apiCallsInprogress > 0
});

const mapDispatchToProps = (dispatch) => ({
  actions: {
    loadShoes: bindActionCreators(shoesActions.loadShoes, dispatch)
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(RootContainer);
