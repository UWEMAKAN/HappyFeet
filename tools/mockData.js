module.exports = {
  shoes: [
    {
      id: 1,
      title: 'Blue Snickers',
      brand: 'Addidas',
      image: './photos/addidas-blue-1-600x600.jpg',
      color: 'Blue',
      price: 325,
      discount: 20,
      rating: 4.4,
    },
    {
      id: 2,
      title: 'Black Snickers',
      brand: 'Others',
      image: './photos/others-black-1-600x600.jpg',
      color: 'Black',
      price: 295,
      discount: 25,
      rating: 4.3,
    },
    {
      id: 3,
      title: 'Black Snickers',
      brand: 'Others',
      image: './photos/others-black-2-600x600.jpg',
      color: 'Black',
      price: 325,
      discount: 25,
      rating: 4.3,
    },
    {
      id: 4,
      title: 'White Snickers',
      brand: 'Nike',
      image: './photos/nike-white-1-600x600.jpg',
      color: 'White',
      price: 450,
      discount: 20,
      rating: 4.6,
    },
    {
      id: 5,
      title: 'Snickers',
      brand: 'Addidas',
      image: './photos/addidas-green-1-600x600.jpg',
      color: 'Green',
      price: 400,
      discount: 25,
      rating: 4.5,
    },
    {
      id: 6,
      title: 'High Hills',
      brand: 'Others',
      image: './photos/others-black-hills-1-600x600.jpg',
      color: 'Black',
      price: 375,
      discount: 25,
      rating: 4.3,
    },
    {
      id: 7,
      title: 'White Snickers',
      brand: 'Nike',
      image: './photos/nike-white-2-600x600.jpg',
      color: 'White',
      price: 450,
      discount: 20,
      rating: 4.6,
    },
    {
      id: 8,
      title: 'Blue Snickers',
      brand: 'Addidas',
      image: './photos/addidas-blue-2-600x600.jpg',
      color: 'Blue',
      price: 400,
      discount: 25,
      rating: 4.5,
    },
    {
      id: 9,
      title: 'Black Leather Shoes',
      brand: 'Others',
      image: './photos/others-black-leather-1-600x600.jpg',
      color: 'Black',
      price: 425,
      discount: 25,
      rating: 4.3,
    },
    {
      id: 10,
      title: 'White Snickers',
      brand: 'Addidas',
      image: './photos/addidas-white-1-600x600.jpeg',
      color: 'White',
      price: 400,
      discount: 25,
      rating: 4.5,
    },
    {
      id: 11,
      title: 'Blue Snickers',
      brand: 'Others',
      image: './photos/others-blue-1-600x600.jpeg',
      color: 'Blue',
      price: 420,
      discount: 25,
      rating: 4.3,
    },
  ]
};
