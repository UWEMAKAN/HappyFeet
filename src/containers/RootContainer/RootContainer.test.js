import React from 'react';
// import fetchMock from 'fetch-mock';
// import {
//   cleanup, fireEvent, render, act
// } from '@testing-library/react';
import { shallow } from 'enzyme';
// import { MemoryRouter } from 'react-router-dom';
import RootContainer, { buyShoeHandler, myShoesHandler } from './RootContainer';
import { shoes } from '../../../tools/mockData';

describe('RootContainer Component', () => {
  test('should render RootContainer Component', () => {
    const wrapper = shallow(<RootContainer />);
    expect.assertions(1);
    expect(wrapper).toMatchSnapshot();
  });

  // describe('Test useEffect', () => {
  //   afterEach(() => {
  //     jest.useRealTimers();
  //   });

  // //   test('should render RootContainer Component', async (done) => {
  // //     fetchMock.mock('*', {
  // //       body: shoes,
  // //       headers: { 'content-type': 'application/json' }
  // //     });
  // //     await act(async () => {
  // //       const container = mount(<MemoryRouter><RootContainer /></MemoryRouter>);
  // //     });
  // //     jest.useFakeTimers();
  // //     jest.runAllImmediates();
  // //   });
  // });

  // test('should render RootContainer Component', (done) => {
  //   global.fetch = fetchMock.mock('*', {
  //     body: shoes,
  //     headers: { 'content-type': 'application/json' }
  //   });
  //   let container;
  //   jest.useFakeTimers();
  //   act(() => {
  //     container = render(<MemoryRouter><RootContainer /></MemoryRouter>);
  //   });
  //   setImmediate(done);
  //   jest.runAllImmediates();
  // });
});

describe('Testing Helpers', () => {
  test('should call setShoe when the function returned by calling buyShoeHandler is execute', () => {
    const setShoe = jest.fn();
    const buyShoe = buyShoeHandler(shoes, setShoe);
    buyShoe(2);
    expect.assertions(1);
    expect(setShoe).toHaveBeenCalledTimes(1);
  });

  test('should call setShoe when the function returned by calling buyShoeHandler is execute', () => {
    const myShoes = [];
    const setMyShoes = jest.fn();
    const setShoe = jest.fn();
    const addToMyShoes = myShoesHandler(myShoes, setMyShoes, setShoe);
    addToMyShoes(shoes[2]);
    expect.assertions(2);
    expect(setShoe).toHaveBeenCalledTimes(1);
    expect(setMyShoes).toHaveBeenCalledTimes(1);
  });
});
