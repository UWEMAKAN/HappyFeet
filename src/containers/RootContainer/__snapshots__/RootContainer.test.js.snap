// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RootContainer Component should render RootContainer Component 1`] = `ShallowWrapper {}`;
